src/aims_ui/api_interaction.py

```diff
@@ -3,43 +3,137 @@
 import requests
 from . import app
 from flask import render_template
+from io import StringIO, BytesIO
 from .models.get_endpoints import get_endpoints
+from .models.get_addresses import get_addresses
 import urllib
+import csv
+
 
 def get_params(all_user_input):
-  """Return a list of aparameters formatted for API header, from class list of inputs"""
-    
+  """Return a list of parameters formatted for API header, from class list of inputs"""
   params = ['verbose=True']
   for param, value in all_user_input.items():
     if not str(value):
       continue
     if (os.getenv('FLASK_ENV') == 'development') and (param == 'epoch'):
       # do not add epoch for testing
       continue
+
+    if type(value) == str:
+      value = value.replace('%', '')
     quoted_param = urllib.parse.quote_plus(str(param))
     quoted_value = urllib.parse.quote_plus(str(value))
-    params.append(quoted_param + '=' + quoted_value)  
-  
+    params.append(quoted_param + '=' + quoted_value)
+
   return '&'.join(params)
 
 
-def api(
-    url,
-    called_from,
-    all_user_input):
+def api(url, called_from, all_user_input):
   """API helper, all pages go through here to interact with API"""
 
-  header = {"Content-Type": "application/json",}
+  header = {
+      "Content-Type": "application/json",
+  }
 
   if (called_from == 'uprn') or (called_from == 'postcode'):
-    url = app.config.get('API_URL') + url + all_user_input.get(called_from)
+    url = app.config.get('API_URL') + url + all_user_input.get(called_from, '')
     params = get_params(all_user_input)
-    r = requests.get(url, params=params, headers=header, )
+    r = requests.get(
+        url,
+        params=params,
+        headers=header,
+    )
 
   elif called_from == 'singlesearch':
     url = app.config.get('API_URL') + url
     proposed_params = {k: v for k, v in all_user_input.items() if v}
     params = get_params(all_user_input)
-    r = requests.get(url, params=params, headers=header, )
+    r = requests.get(
+        url,
+        params=params,
+        headers=header,
+    )
 
   return r
+
+
+def multiple_address_match(file, all_user_input, app, download=False):
+  final_csv = 'id, inputAddress, matchedAddress, uprn, matchType, confidenceScore, documentScore, rank\n\r'
+  ths = [{'value': x, 'ariaSort': None} for x in final_csv.split(',')]
+  trs = []
+
+  contents = file.readlines()
+  if download:
+    proxy = StringIO()
+    writer = csv.writer(proxy)
+    writer.writerow(final_csv.split(','))
+
+  line_count = 0
```

Comment:
If this line count is directly tracking the number of iterations, consider using `for line_count, line in enumerate(contents)` instead, and removing the increment later on.

Reply:
This counter counts each match for each line (with multiple matches or sinle matches per line). 
Got 27 with this method, which is the number of lines inputted by the initial CSV
Should be 108, the number of outputted lines for the final CSV